versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/persist/Persist.java

```diff
@@ -0,0 +1,341 @@
+   *
+   * <p><em>Do not use this function from service implementations, use {@link ReferenceLogic}
+   * instead!</em>
+   *
+   * @return the updated {@link Reference}, if the reference exists, is not marked as {@link
+   *     Reference#deleted() deleted} and the {@link Reference#pointer()} update succeeded.
+   * @throws RefNotFoundException if a reference with the same name does not exist
+   * @throws RefConditionFailedException if the existing reference is marked as deleted or its
+   *     pointer is different
+   */
+  @Nonnull
+  @jakarta.annotation.Nonnull
+  Reference updateReferencePointer(
+      @Nonnull @jakarta.annotation.Nonnull Reference reference,
+      @Nonnull @jakarta.annotation.Nonnull ObjId newPointer)
+      throws RefNotFoundException, RefConditionFailedException;
+
+  /**
+   * Low-level, find a reference.
+   *
+   * <p><em>Do not use this function from service implementations, use {@link ReferenceLogic}
+   * instead!</em>
+   *
+   * @return the reference or {@code null}, if it does not exist
+   */
+  Reference findReference(@Nonnull @jakarta.annotation.Nonnull String name);
+
+  /**
+   * Like {@link #findReference(String)}, but finds multiple references by name at once, leveraging
+   * bulk queries against databases.
+   *
+   * <p>Non-existing references are returned as {@code null} elements in the returned array.
+   *
+   * <p><em>Do not use this function from service implementations, use {@link ReferenceLogic}
+   * instead!</em>
+   */
+  @Nonnull
+  @jakarta.annotation.Nonnull
+  Reference[] findReferences(@Nonnull @jakarta.annotation.Nonnull String[] names);
+
+  // Objects
+
+  /**
+   * Fetches a single object.
+   *
+   * <p>If the type of the requested object is known, prefer {@link #fetchTypedObj(ObjId, ObjType,
+   * Class)}.
+   *
+   * @return The returned object will be a concrete type according to its {@link ObjType},
+   *     (de)serialization is handled by the database specific implementation of {@link Persist}.
+   *     Returns {@code null} for {@code id} equal to {@link ObjId#EMPTY_OBJ_ID}
+   * @throws ObjNotFoundException with the ID for which no {@link Obj objects} exist
+   * @see #fetchObjType(ObjId)
+   * @see #fetchTypedObj(ObjId, ObjType, Class)
+   * @see #fetchObjs(ObjId[])
+   */
+  Obj fetchObj(@Nonnull @jakarta.annotation.Nonnull ObjId id) throws ObjNotFoundException;
+
+  /**
+   * Retrieves the object with ID {@code id}, having the same {@link ObjType type}.
+   *
+   * @return the object with the requested type, {@code null} for {@link ObjId#EMPTY_OBJ_ID}
+   * @throws ObjNotFoundException with the ID for which no matching {@link Obj objects} exist,
+   *     either the no object with the given ID exists or that object is not of the requested type
+   * @see #fetchObjType(ObjId)
+   * @see #fetchObj(ObjId)
+   * @see #fetchObjs(ObjId[])
+   */
+  <T extends Obj> T fetchTypedObj(
+      @Nonnull @jakarta.annotation.Nonnull ObjId id, ObjType type, Class<T> typeClass)
```

Comment:
Neccessary for type-safety / generics

Reply:
Why not make it a type parameter on `ObjType`? smth like: ` <T extends Obj> T fetchTypedObj(..., ObjType<T> type)`

Reply:
IMHO, having a `typeClass` as a parameter that is passed in, but not used, does not add type safety, because the caller can use any class as `typeClass` and still get a `ClassCastException` in runtime :thinking:

Reply:
It's a Java language necessity.

Reply:
`ObjType` could be a generic class or itnerface. It does not have to be an `enum`, I guess.

Reply:
We can iron this out later, of course.